daemon/src/contract.rs

```diff
@@ -0,0 +1,309 @@
+///
+/// The refund timelock is important in case the oracle disappears or never publishes a signature.
+/// Ideally, both users collaboratively settle in the refund scenario. This factor is important if
+/// the users do not settle collaboratively.
+/// `1.5` times the term as defined in CFD order should be safe in the extreme case where a user
+/// publishes the commit transaction right after the contract was initialized. In this case, the
+/// oracle still has `1.0 * cfdorder.term` time to attest and no one can publish the refund
+/// transaction.
+/// The downside is that if the oracle disappears: the users would only notice at the end
+/// of the cfd term. In this case the users has to wait for another `1.5` times of the
+/// term to get his funds back.
+pub const REFUND_THRESHOLD: f32 = 1.5;
+
+#[derive(Clone)]
+#[allow(dead_code)]
+pub enum OwnParams {
+    Maker(PartyParams),
+    Taker(PartyParams),
+}
+
+enum Role {
+    Maker,
+    Taker,
+}
+
+/// A convenience struct for storing PartyParams and PunishParams for both
+/// parties and the
+struct AllParams {
+    pub own: PartyParams,
+    pub own_punish: PunishParams,
+    pub other: PartyParams,
+    pub other_punish: PunishParams,
+    pub own_role: Role,
+}
+
+impl AllParams {
+    fn new(
+        own: PartyParams,
+        own_punish: PunishParams,
+        other: PartyParams,
+        other_punish: PunishParams,
+        own_role: Role,
+    ) -> Self {
+        Self {
+            own,
+            own_punish,
+            other,
+            other_punish,
+            own_role,
+        }
+    }
+
+    fn maker(&self) -> &PartyParams {
+        match self.own_role {
+            Role::Maker => &self.own,
+            Role::Taker => &self.other,
+        }
+    }
+
+    fn taker(&self) -> &PartyParams {
+        match self.own_role {
+            Role::Maker => &self.other,
+            Role::Taker => &self.own,
+        }
+    }
+
+    fn maker_punish(&self) -> &PunishParams {
+        match self.own_role {
+            Role::Maker => &self.own_punish,
+            Role::Taker => &self.other_punish,
+        }
+    }
+    fn taker_punish(&self) -> &PunishParams {
+        match self.own_role {
+            Role::Maker => &self.other_punish,
+            Role::Taker => &self.own_punish,
+        }
+    }
+}
+
+/// Given an initial set of parameters, sets up the CFD contract with the other party.
+/// Passing OwnParams identifies whether caller is the maker or the taker.
+///
+/// Returns the [`FinalizedCfd`] which contains the lock transaction, ready to be signed and sent to
+/// the other party. Signing of the lock transaction is not included in this function because we
+/// want the Cfd actor to own the wallet.
+pub fn setup_contract(
+    send_to_other: impl Fn(SetupMsg),
+    own_params: OwnParams,
+    sk: SecretKey,
+    oracle_pk: schnorrsig::PublicKey,
+    order: Order,
+) -> (
+    impl Future<Output = FinalizedCfd>,
+    mpsc::UnboundedSender<SetupMsg>,
+) {
```

Comment:
As a general theme, try to put the public API of a module at the very top and order things from highest to lowest level of importance. That way, as you start reading the code, you know what this module does and by continuing to read, you will learn about the details how it is implemented. Assuming descriptive names, just reading the top-level public function may give you enough information about what this is doing without having to scan down further!

Reply:
that makes a lot of sense! will change things around to reflect this.

Reply:
@thomaseizinger I've read somewhere (in the rustbook?) that it's customary for `const` to  be at the top of the file - shall I keep the `REFUND_TRESHOLD` above the `fn new()` ?

Reply:
Yeah consts can stay at the top :)